### PR TITLE
Add support for elevation in DrawerLayoutAndroid

### LINKED ISSUE #6022: Support elevation for the drawer in DrawerLayoutAndroid
`style` on `View` supports an `elevation` property on Android 5+ that works with Android's Elevation API.

But there is no external access to the view for the actual drawer used in `DrawerLayoutAndroid`. It can't be replaced like `navigationBar` in `Navigation` allows, setting `style={{elevation: 16}}` doesn't work, and there's no prop to pass styles for the drawer. `renderNavigationView` exists, but that defines the contents of the drawer, not the drawer itself, so elevation on it doesn't work.

Under the Material Design guidelines navigation drawers should have an elevation of 16.
https://www.google.com/design/spec/what-is-material/elevation-shadows.html#elevation-shadows-elevation-android-

On the navigation drawers page you can see the shadow cast by the elevation:
https://www.google.com/design/spec/patterns/navigation-drawer.html

### PR DESCRIPTION
**Pull request by janicduplessis**
It didn't work for a few reason. First, the drawer view NEEDS to have a background color or no shadow will ever render. Second, we need to use the `setDrawerElevation` method instead of `setElevation` for DrawerLayout. Finally we need to actually pass the style value (maybe we could just pass elevation but I don't really think it can cause any issues) down to the native component as it is not the case at the moment.

I also added a default style to elevation of 16 which is the standard for material design according to https://www.google.com/design/spec/patterns/navigation-drawer.html#navigation-drawer-specs. I could also default it to 0 so it keeps the same appearance as before but I think it looks better this way.

Closes #6022
**Test plan**
Tested using the DrawerLayout in the UIExplorer app.

Before, elevation 0
screen shot 2016-02-23 at 1 55 42 am

After, elevation 16
screen shot 2016-02-23 at 1 54 48 am

### CODE REVIEW

**Comment by bestander** [comment]
FYI, BUCK is not happy with the change.
https://circleci.com/gh/facebook/react-native/2971

**Comment by janicduplessis** [comment]
```
Module com.facebook.react.views.drawer.ReactDrawerLayoutManager has already registered a property named "elevation"
```

The props preprocessor doesn't like me overriding the elevation prop implementation :(

@lexs would this be easy to support? Otherwise I can just rename the prop to something else and set it in the js component.

**Comment by lexs** [comment]
Assuming they have the same type etc you should be able to override the
method.

**Comment by janicduplessis** [comment]
I tried adding the @Override annotation to see if it changed anything but it didn't. One has a generic type maybe it's why it doesn't work.
Base class:
```
public void setElevation(T view, float elevation) {
```
Subclass
```
public void setElevation(ReactDrawerLayout view, float elevation) {
```

**Comment by lexs** [comment]
@janicduplessis Which version are you running on? I know there was a bug regarding overriding methods previously. I'll check tomorrow what version the fix went into.

**Comment by janicduplessis** [comment]
@lexs I assume you mean what version of buck, I just tried with a fresh pull from master. That is what the build running on CI is doing too.

File: Libraries/Components/DrawerAndroid/DrawerLayoutAndroid.android.js [hunk 1]
```diff
@@ -216,6 +217,7 @@ var styles = StyleSheet.create({
     position: 'absolute',
     top: 0,
     bottom: 0,
+    backgroundColor: 'white',
```
Comment (mkonicek) [inline Libraries/Components/DrawerAndroid/DrawerLayoutAndroid.android.js]: Is this relevant to this PR (adding support for elevation)?

**Comment by mkonicek** [comment]
Not sure we want to set the color to white, besides that looks good overall. Thanks for doing this! Once you and @lexs figure out the tests let's merge it.

File: Libraries/Components/DrawerAndroid/DrawerLayoutAndroid.android.js (hunk 1 above)
Comment (janicduplessis) [inline Libraries/Components/DrawerAndroid/DrawerLayoutAndroid.android.js]: Yes the shadow doesn't draw if the view doesn't have a backgroundColor. I don't think it will ever be visible unless someone doesn't set any background to their drawer view in their app.

**Comment by janicduplessis** [comment]
@lexs Any update on this? If it seems hard to fix I can just rename the prop for now to unblock this.

**Comment by mkonicek** [comment]
@janicduplessis Can you please rebase to see both Travis (iOS & JS) and CircleCI (Android) tests passing?

We've fixed the tests so we're back to green now:
https://travis-ci.org/facebook/react-native/builds
https://circleci.com/gh/facebook/react-native/tree/master

**Comment by lexs** [comment]
@janicduplessis Sorry for not responding sooner, totally missed the emails.

You should be able to have the following:
```java
  @Override
  public void setElevation(ReactDrawerLayout view, float elevation) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      view.setDrawerElevation(PixelUtil.toPixelFromDIP(elevation));
    }
  }
```

**Comment by janicduplessis** [comment]
@lexs I just decided to check how to fix it (that timing!), I changed the ReactPropertyProcessor to check types and only throw if the type are different. This allow leaving the annotation like I did initially. Let me know if you thinks that's fine otherwise I can do what you suggested. :)

**Comment by lexs** [comment]
Yeah you'll need to change it to the `@Override` I mentioned. 

The reason for this is that while the annotation processor might behave correctly in the case where you add the same property again the old, reflection-based implementation will not (and it's still used if you don't build by Buck).

I agree that you should be able to override properties in the way you did initially but currently behavior is basically undefined. The reflection-based version will pick one of them. This is why I added the error initially. In practice the override approach has worked every time this have come up (two times now). So fixing this is pretty low-prio.

**Comment by janicduplessis** [comment]
Yea it's a pretty rare case, it did work when I tested with the reflection based version (it picked the good one I guess). I changed the message a bit to point to the right direction if it happens again and changed it to what you suggested.

**Comment by janicduplessis** [comment]
@mkonicek All the tests are passing now and the prop issue is figured out!

**Comment by lexs** [comment]
@facebook-github-bot shipit

**Comment by lexs** [comment]
Thanks!

**Comment by mkonicek** [comment]
Fails to compile in fb CI with:

    ReactAndroid/src/main/java/com/facebook/react/views/drawer/ReactDrawerLayoutManager.java:96:      
    error: cannot find symbol
    [CONTEXT]       view.setDrawerElevation(PixelUtil.toPixelFromDIP(elevation));

Any ideas?

**Comment by lexs** [comment]
Yes, sorry for not getting back sooner. This fails internally because we're on an older version of the Android support lib. We've landed a newer version and I'm gonna try to switch over the `DrawerAndroid` component to it. It shouldn't need any changes on your part.

**Comment by lexs** [comment]
@janicduplessis Turns out this will be really hard to land as-is. The reason is we're on an older version of the support library internally an upgrading is going to take a while. I'm sorry about this. To land this for now could you perhaps rewrite the call to `DrawerLayout`.setDrawerElevation() to use reflection (or some other method that doesn't require the method to exist) :).

**Comment by janicduplessis** [comment]
@lexs Updated to call the method with reflection as you suggested!

File: ReactAndroid/src/main/java/com/facebook/react/views/drawer/ReactDrawerLayoutManager.java [hunk 2]
```diff
@@ -90,6 +93,21 @@ public void setDrawerLockMode(ReactDrawerLayout view, @Nullable String drawerLoc
...
+  public void setElevation(ReactDrawerLayout view, float elevation) {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+      // Facebook is using an older version of the support lib internally that doesn't support
+      // setDrawerElevation so we invoke it using reflection.
+      // TODO: Call the method directly when this is no longer needed.
+      try {
+        Method method = ReactDrawerLayout.class.getMethod("setDrawerElevation", float.class);
+        method.invoke(view, PixelUtil.toPixelFromDIP(elevation));
+      } catch (Exception e) {
+        // Ignore errors when the method is not present in older versions of support lib.
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/drawer/ReactDrawerLayoutManager.java]: Sounds good, maybe just `Flog.d` so people can debug why the method doesn't work for them?

Does `getMethod` return `null`? In that case the NPE won't be very user friendly. If so might be worth checking if the method exists and logging a clear message.

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/drawer/ReactDrawerLayoutManager.java]: I wasn't sure how to handle the error here but I don't think it can happen for users outside facebook since the gradle build already has a dependency on a recent version of the support lib that always has the method so I'm not sure if logging something is needed.

`getMethod` cannot return null, it will throw if it cannot find the method or something else is wrong.

Comment (lexs) [inline ReactAndroid/src/main/java/com/facebook/react/views/drawer/ReactDrawerLayoutManager.java]: Can you add a warning for the error case (`FLog.e`). Somebody might try to use this method internally and be suprised about it not working.

**Comment by janicduplessis** [comment]
Updated with the warning log.

File: ReactAndroid/src/main/java/com/facebook/react/views/drawer/ReactDrawerLayoutManager.java (hunk 2 above)
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/drawer/ReactDrawerLayoutManager.java]: Good point, logging an error is better than just debug in this case.

**Comment by mkonicek** [comment]
@facebook-github-bot shipit

**Comment by mkonicek** [comment]
@facebook-github-bot shipit

**Comment by mkonicek** [comment]
@lexs I see you abandoned the internal diff with a comment "will reimport". Since it uses reflection now I'll merge it as is.

**Comment by lexs** [comment]
Yes, that was my intention. Thanks Martin!
